Compatibility overhaul

TexKiller opened the pull request:
<h2>Table of Contents</h2>

- <a href="#user-content-introduction">Introduction</a>
- <a href="#user-content-why-is-firefox-harder">Why is Firefox harder</a>
- <a href="#user-content-compatibility-table">Compatibility table</a>
- <a href="#user-content-changes-to-the-code">Changes to the code</a>
  - <a href="#user-content-streamsaverjs">StreamSaver.js</a>
  - <a href="#user-content-swjs">sw.js</a>
  - <a href="#user-content-mitmhtml">mitm.html</a>
  - <a href="#user-content-pinghtml">ping.html</a>
  - <a href="#user-content-pingjs">ping.js</a>
- <a href="#user-content-changes-to-functionality">Changes to functionality</a>
  - <a href="#user-content-download-from-iframe">Download from <code>&lt;iframe&gt;</code></a>
  - <a href="#user-content-pinging-on-chromeopera">Pinging on Chrome/Opera</a>
  - <a href="#user-content-web-extensions-background-page-on-chromeopera">Web Extension's background page on Chrome/Opera</a>
  - <a href="#user-content-http-pages-on-firefox">HTTP pages on Firefox</a>
  - <a href="#user-content-web-extensions-on-firefox">Web Extensions on Firefox</a>
- <a href="#user-content-examples-for-testing">Examples for testing</a>
  - <a href="#user-content-web-extensions-background-page-download">Web Extension's background page download</a>
  - <a href="#user-content-web-extensions-page-on-a-tab-download">Web Extension's page on a tab download</a>
  - <a href="#user-content-web-page-on-http">Web page on HTTP</a>
  - <a href="#user-content-web-page-on-https">Web page on HTTPS</a>
- <a href="#user-content-things-to-look-forward-to">Things to look forward to</a>
  - <a href="#user-content-transferable-streams">Transferable Streams</a>
  - <a href="#user-content-starting-service-worker-from-moz-extension">Starting Service Worker from <code>moz-extension://*</code></a>

<h2><a name="introduction">Introduction</a></h2>

This PR is meant to introduce full support to Firefox in all contexts while maintaining (and lightly improving) the existing Chrome/Opera support.
In the process of making this I have discovered and extensively tested many methods of doing each step StreamSaver.js does in all contexts/browsers, and have isolated the best method for each.

I am considering four StreamSaver.js steps:
1) Starting the Service Worker
2) Sending message port to the Service Worker
3) Starting the download
4) Keeping the Service Worker alive

**PS:** Step 4 should be unnecessary if <a href="https://github.com/whatwg/streams/blob/master/transferable-streams-explainer.md">Transferable Streams</a> are supported. In step 2, instead of sending only the message port, we would also send the stream.

I am also considering four contexts:
1) Background page of a Web Extension
2) Web Extension's page loaded in a tab
3) HTTP page
4) HTTPS page

<h2><a name="why-is-firefox-harder">Why is Firefox harder</a></h2>

Firefox requires some stronger methods mainly because of the following reasons:
- Web Extensions on Firefox run on their own processes, which means they can't interact with Service Workers that are running on regular tab processes (and vice versa);
- Web Extensions can't ([yet](https://bugzilla.mozilla.org/show_bug.cgi?id=1344561)) start Service Workers from their own <code>moz-extension://*</code> URLs on Firefox - they can only start Service Workers from external HTTPS sites;
- Web Extension's background pages on Firefox are incapable of starting downloads without using a tab/window;
- Web Extension's background pages on Firefox are incapable of opening popups using <code><a href="https://developer.mozilla.org/docs/Web/API/Window/open">window.open</a></code>;
- Firefox doesn't have a bug that Chrome and Opera have. This bug allows a closed popup to keep a Service Worker alive forever (or until the entire browser is closed).

The first two reasons imply that Web Extensions on Firefox have to either start the Service Worker from a popup (outside of the Web Extension's process, since it will be from a tab with an external HTTPS site) or from the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> (inside the Web Extension's process).

But since the download link will always be from an outside HTTPS URL, if you open it on a popup it will be loaded from outside of the Web Extension's process.
If the Service Worker was started from inside the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code>, it won't be available for that tab's process, and the 404 page will be shown.

However, if you load the download link inside another <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code>, it will still be inside the Web Extension's process, and it will have the Service Worker available.
That way we can have tabs with Web Extension's <code>moz-extension://*</code> pages start both the Service Worker and the download without opening any popups.
Too bad the background page can't start a download from the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> on Firefox :persevere:

On a side note, if it were possible to start a Service Worker from <code>moz-extension://*</code> URLs, if a download URL from that protocol were loaded on a tab then the Web Extension's process would be used, and it would have access to the Service Worker started from the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code>.
But for that the [1344561 bug on bugzilla](https://bugzilla.mozilla.org/show_bug.cgi?id=1344561) has to be fixed first.

The fourth reason means popups have to be opened from the background page using Web Extension exclusive functions, like <code><a href="https://developer.chrome.com/extensions/tabs#method-create">chrome.tabs.create</a></code>.
Those functions don't provide an easy way to pass a message port to the popup, and the only way I could find of doing that was through a Shared Worker started from the popup iself.

Last but not least, since the closed popup can't keep the Service Worker alive forever on Firefox, we need another way of doing that. We can't ping the Service Worker from HTTP pages (because of security restrictions) or from Web Extension's pages if the Service Worker is started from a popup (because of different processes).
For those situations the only way I could find to ping the Service Worker without keeping a popup open was through a Shared Worker.

<h2><a name="compatibility-table">Compatibility table</a></h2>

I know this is a lot of text. In an attempt to facilitate the understanding of everything this PR does and why, I have made a (very ugly :disappointed_relieved:) table with collapsable cells.
You can click on a cell to read an explanation of what the method does and why the other methods are worse than the chosen one (click on them to see the reason).

<table>
<tr><th rowspan="6"></th></tr>
<tr><th colspan="4"><h3>Firefox</h3></th><th colspan="4"><h3>Chrome/Opera</h3></th></tr>
<tr></tr>
<tr><th colspan="2">Web Extension</th><th colspan="2">Web</th><th colspan="2">Web Extension</th><th colspan="2">Web</th></tr>
<tr></tr>
<tr><th>Back- ground Page</th><th>Tab</th><th>HTTP</th><th>HTTPS</th><th>Back- ground Page</th><th>Tab</th><th>HTTP</th><th>HTTPS</th></tr>
<tr></tr>
<tr><th><h6>Start<br />Service<br />Worker</h6></th>
<td><h6><details><summary>

```js
'From
chrome
.tabs.
create
```
</summary>
<kbd>Create a new tab with <code>mitm.html</code> from a domain on HTTPS using the <code><a href="https://developer.chrome.com/extensions/tabs#method-create">chrome.tabs.create</a></code> function.</kbd>
<h3>Worse alternatives:</h3>
<details><summary>

```php
 From
iframe
```
</summary>
<kbd>It is <b>possible</b> (using <code>mitm.html</code> from a domain on HTTPS outside of the Web Extension), but the Service Worker is trapped inside the extension's process and then it is impossible to start the download, at least without an ugly <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> inside a popup, which requires loading twice before the download starts (once for the popup page, and again for the download link inside the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code>).<br />
<br />
One advantage of this, however, is that it opens only one popup instead of two, but the same can be accomplished if we reuse the same port between downloads (maybe in a future PR).</kbd>
</details>
<details><summary>

```nginx
 From
window
.open
```
</summary>
<kbd>Web Extensions on Firefox cannot open popup windows using <code><a href="https://developer.mozilla.org/docs/Web/API/Window/open">window.open</a></code> from the background page.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
 From
iframe
```
</summary>
<kbd>Open <code>mitm.html</code> from a domain on HTTPS inside an <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code>.</kbd>
<h3>Worse alternatives:</h3>
<details><summary>

```js
'
 From
chrome
.tabs.
create
```
</summary>
<kbd>It is <b>possible</b> (using <code>mitm.html</code> from a domain on HTTPS outside of the Web Extension), but it opens popups, which is not necessary.<br />
<br />
One small advantage of this, however, is that it doesn't require starting the download from the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code>.</kbd>
</details>
<details><summary>

```nginx
 From
window
.open
```
</summary>
<kbd>It is <b>possible</b> (using <code>mitm.html</code> from a domain on HTTPS outside of the Web Extension), but it requires user interaction <b>and</b> opens popups, both of which are not necessary.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```nginx
 From
window
.open
```
</summary>
<kbd>Open <code>mitm.html</code> from a domain on HTTPS in a popup using the <code><a href="https://developer.mozilla.org/docs/Web/API/Window/open">window.open</a></code> function.</kbd>
<h3>Worse alternatives:</h3>
<details><summary>

```js
'From
chrome
.tabs.
create
```
</summary>
<kbd>The <code><a href="https://developer.chrome.com/extensions/tabs#method-create">chrome.tabs.create</a></code> function is only available for Web Extensions, and the chosen <code><a href="https://developer.mozilla.org/docs/Web/API/Window/open">window.open</a></code> function facilitates communication with the <code>mitm.html</code> page.</kbd>
</details>
<details><summary>

```php
 From
iframe
```
</summary>
<kbd>Pages loaded inside <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> on HTTP pages are incapable of interacting with Service Workers.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
 From
iframe
```
</summary>
<kbd>Open <code>mitm.html</code> inside an <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code>.</kbd>
<h3>Worse alternatives:</h3>
<details><summary>

```js
'From
chrome
.tabs.
create
```
</summary>
<kbd>The <code><a href="https://developer.chrome.com/extensions/tabs#method-create">chrome.tabs.create</a></code> function is only available for Web Extensions, and it opens popups, which is not necessary.</kbd>
</details>
<details><summary>

```nginx
 From
window
.open
```
</summary>
<kbd>It is <b>possible</b>, but it requires user interaction <b>and</b> opens popups, both of which are not necessary.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
 From
iframe
```
</summary>
<kbd>Open <code>mitm.html</code> inside an <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code>.</kbd>
<h3>Worse alternatives:</h3>
<details><summary>

```js
'From
chrome
.tabs.
create
```
</summary>
<kbd>It is <b>possible</b>, but it opens popups, which is not necessary.</kbd>
</details>
<details><summary>

```nginx
 From
window
.open
```
</summary>
<kbd>It is <b>possible</b>, but it opens popups, which is not necessary.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
 From
iframe
```
</summary>
<kbd>Open <code>mitm.html</code> inside an <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code>.</kbd>
<h3>Worse alternatives:</h3>
<details><summary>

```js
'From
chrome
.tabs.
create
```
</summary>
<kbd>It is <b>possible</b>, but it opens popups, which is not necessary.</kbd>
</details>
<details><summary>

```nginx
 From
window
.open
```
</summary>
<kbd>It is <b>possible</b>, but it requires user interaction <b>and</b> opens popups, both of which are not necessary.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```nginx
 From
window
.open
```
</summary>
<kbd>Open <code>mitm.html</code> in a popup using the <code><a href="https://developer.mozilla.org/docs/Web/API/Window/open">window.open</a></code> function.</kbd>
<h3>Worse alternatives:</h3>
<details><summary>

```js
'From
chrome
.tabs.
create
```
</summary>
<kbd>The <code><a href="https://developer.chrome.com/extensions/tabs#method-create">chrome.tabs.create</a></code> function is only available for Web Extensions, and the chosen <code><a href="https://developer.mozilla.org/docs/Web/API/Window/open">window.open</a></code> function facilitates communication with the <code>mitm.html</code> page.</kbd>
</details>
<details><summary>

```php
 From
iframe
```
</summary>
<kbd>Pages loaded inside <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> on HTTP pages are incapable of interacting with Service Workers.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
 From
iframe
```
</summary>
<kbd>Open <code>mitm.html</code> inside an <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code>.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
'
 From
chrome
.tabs.
create
```
</summary>
<kbd>The <code>chrome .tabs.create</code> function is only available for Web Extensions, and it opens popups, which is not necessary.</kbd>
</details>
<details><summary>

```nginx
 From
window
.open
```
</summary>
<kbd>It is <b>possible</b>, but it requires user interaction <b>and</b> opens popups, both of which are not necessary.</kbd>
</details>
</details></h6></td></tr>
<tr></tr>
<tr><th><h6>Send messa- ge port</h6></th>
<td><h6><details><summary>

```js
 'By
Shared
Worker
```
</summary>
<kbd>Send the message port through a Shared Worker that was first created by the popup and is then opened inside an <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> on the background page.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```php
  By
iframe
```
</summary>
<kbd>It would be <b>possible</b> if the Service Worker were on the Web Extension's process, but that would complicate the download. See the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> alternative to starting the Service Worker for more information.</kbd>
</details>
<details><summary>

```nginx
  By
popup
```
</summary>
<kbd>It is not possible, since the background page cannot open popups using <code><a href="https://developer.mozilla.org/docs/Web/API/Window/open">window.open</a></code> on Firefox and the tabs created with <code><a href="https://developer.chrome.com/extensions/tabs#method-create">chrome.tabs.create</a></code> have no way of receiving message ports directly.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
  By
iframe
```
</summary>
<kbd>Send the message port through <code><a href="https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage">iframe.contentWindow.postMessage</a></code>.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
 'By
Shared
Worker
```
</summary>
<kbd>It is <b>possible</b>, but not necessary.</kbd>
</details>
<details><summary>

```nginx
 By
popup
```
</summary>
<kbd>It is <b>possible</b> if a popup is first created with <code><a href="https://developer.mozilla.org/docs/Web/API/Window/open">window.open</a></code> to start the Service Worker, but doing that is not necessary.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```nginx
 By
popup
```
</summary>
<kbd>Send the message port through <code><a href="https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage">popup.postMessage</a></code>.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
 'By
Shared
Worker
```
</summary>
<kbd>It is <b>possible</b>, but not necessary.</kbd>
</details>
<details><summary>

```php
  By
iframe
```
</summary>
<kbd>It is not possible. See the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> alternative to start the Service Worker for more information.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
  By
iframe
```
</summary>
<kbd>Send the message port through <code><a href="https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage">iframe.contentWindow.postMessage</a></code>.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
 'By
Shared
Worker
```
</summary>
<kbd>It is <b>possible</b>, but not necessary.</kbd>
</details>
<details><summary>

```nginx
 By
popup
```
</summary>
<kbd>It is <b>possible</b> if a popup is first created with <code><a href="https://developer.mozilla.org/docs/Web/API/Window/open">window.open</a></code> to start the Service Worker, but doing that is not necessary.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
  By
iframe
```
</summary>
<kbd>Send the message port through <code><a href="https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage">iframe.contentWindow.postMessage</a></code>.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
 'By
Shared
Worker
```
</summary>
<kbd>It is <b>possible</b>, but not necessary.</kbd>
</details>
<details><summary>

```nginx
 By
popup
```
</summary>
<kbd>It is <b>possible</b> if a popup is first created with <code><a href="https://developer.mozilla.org/docs/Web/API/Window/open">window.open</a></code> to start the Service Worker, but doing that is not necessary.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
  By
iframe
```
</summary>
<kbd>Send the message port through <code><a href="https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage">iframe.contentWindow.postMessage</a></code>.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
 'By
Shared
Worker
```
</summary>
<kbd>It is <b>possible</b>, but not necessary.</kbd>
</details>
<details><summary>

```nginx
 By
popup
```
</summary>
<kbd>It is <b>possible</b> if a popup is first created with <code><a href="https://developer.mozilla.org/docs/Web/API/Window/open">window.open</a></code> to start the Service Worker, but doing that is not necessary.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```nginx
 By
popup
```
</summary>
<kbd>Send the message port through <code><a href="https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage">popup.postMessage</a></code>.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
 'By
Shared
Worker
```
</summary>
<kbd>It is <b>possible</b>, but not necessary.</kbd>
</details>
<details><summary>

```php
  By
iframe
```
</summary>
<kbd>It is not possible. See the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> alternative to start the Service Worker for more information.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
  By
iframe
```
</summary>
<kbd>Send the message port through <code>iframe .content Window .post Message</code>.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
 'By
Shared
Worker
```
</summary>
<kbd>It is <b>possible</b>, but not necessary.</kbd>
</details>
<details><summary>

```nginx
 By
popup
```
</summary>
<kbd>It is <b>possible</b> if a popup is first created with <code>window .open</code> to start the Service Worker, but doing that is not necessary.</kbd>
</details>
</details></h6></td></tr>
<tr></tr>
<tr><th><h6>Start down- load</h6></th>
<td><h6><details><summary>

```js
 'On
chrome
.tabs.
create
```
</summary>
<kbd>Create a new tab with the download link using the <code><a href="https://developer.chrome.com/extensions/tabs#method-create">chrome.tabs.create</a></code> function, and remove it using the <code>chrome.tabs.remove</code> function after the download starts.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```php
  On
iframe
```
</summary>
<kbd>As far as I know it is not possible to start downloads from the background page on Firefox without creating a tab.<br />
<br />
It would be <b>possible</b> to have the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> inside a tab created with <code><a href="https://developer.chrome.com/extensions/tabs#method-create">chrome.tabs.create</a></code>, but that is bad. See the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> alternative to start the Service Worker for more information.</kbd>
</details>
<details><summary>

```nginx
  On
window
.open
```
</summary>
<kbd>Web Extensions on Firefox cannot open popup windows using <code><a href="https://developer.mozilla.org/docs/Web/API/Window/open">window.open</a></code> from the background page.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
  On
iframe
```
</summary>
<kbd>Create a hidden <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> with the download link, and remove it after the download starts.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```nginx
  On
window
.open
```
</summary>
<kbd>It would be <b>possible</b> if the Service Worker were started from a popup, but opening popups is not necessary.</kbd>
</details>
<details><summary>

```js
 'On
chrome
.tabs.
create
```
</summary>
<kbd>It would be <b>possible</b> if the Service Worker were started from a popup, but opening popups is not necessary.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
  On
iframe
```
</summary>
<kbd>Create a hidden <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> with the download link, and remove it after the download starts.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
 'On
chrome
.tabs.
create
```
</summary>
<kbd>The <code><a href="https://developer.chrome.com/extensions/tabs#method-create">chrome.tabs.create</a></code> function is only available for Web Extensions, and opening a popup is not necessary.</kbd>
</details>
<details><summary>

```nginx
  On
window
.open
```
</summary>
<kbd>It is <b>possible</b>, but opening a popup is not necessary.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
  On
iframe
```
</summary>
<kbd>Create a hidden <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> with the download link, and remove it after the download starts.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
 'On
chrome
.tabs.
create
```
</summary>
<kbd>The <code><a href="https://developer.chrome.com/extensions/tabs#method-create">chrome.tabs.create</a></code> function is only available for Web Extensions, and opening a popup is not necessary.</kbd>
</details>
<details><summary>

```nginx
  On
window
.open
```
</summary>
<kbd>It is <b>possible</b>, but opening popups is not necessary.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
  On
iframe
```
</summary>
<kbd>Create a hidden <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> with the download link, and remove it after the download starts.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
 'On
chrome
.tabs.
create
```
</summary>
<kbd>It is <b>possible</b>, but not necessary.</kbd>
</details>
<details><summary>

```nginx
  On
window
.open
```
</summary>
<kbd>It is <b>possible</b>, but opening popups is not necessary.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
  On
iframe
```
</summary>
<kbd>Create a hidden <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> with the download link, and remove it after the download starts.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
 'On
chrome
.tabs.
create
```
</summary>
<kbd>It is <b>possible</b>, but not necessary.</kbd>
</details>
<details><summary>

```nginx
  On
window
.open
```
</summary>
<kbd>It is <b>possible</b>, but opening popups is not necessary.</kbd>
</details></h6></td>
<td><h6><details><summary>

```nginx
  On
window
.open
```
</summary>
<kbd>Open a popup with the download link using the <code><a href="https://developer.mozilla.org/docs/Web/API/Window/open">window.open</a></code> function, and close it after the download starts.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
 'On
chrome
.tabs.
create
```
</summary>
<kbd>The <code><a href="https://developer.chrome.com/extensions/tabs#method-create">chrome.tabs.create</a></code> function is only available for Web Extensions, and opening a popup is not necessary.</kbd>
</details>
<details><summary>

```php
  On
iframe
```
</summary>
<kbd>Chrome and Opera won't let Service Workers handle loading of pages inside the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> on insecure pages, so the download can't be started this way.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
  On
iframe
```
</summary>
<kbd>Create a hidden <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> with the download link, and remove it after the download starts.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
 'On
chrome
.tabs.
create
```
</summary>
<kbd>The <code>chrome .tabs.create</code> function is only available for Web Extensions, and opening a popup is not necessary.</kbd>
</details>
<details><summary>

```nginx
  On
window
.open
```
</summary>
<kbd>It is <b>possible</b>, but opening popups is not necessary.</kbd>
</details>
</details></h6></td></tr>
<tr></tr>
<tr><th><h6>Keep Service Worker alive</h6></th>
<td><h6><details><summary>

```js
'Ping
 from 
Shared
Worker
''*
```
</summary>
<kbd>Ping the Service Worker every 29 seconds from inside the Shared Worker.<br /><b>* The Shared Worker must have been started by the popup, otherwise it will reside in the Web Extension's process and it wont work.</b></kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```php
 Ping
 from
iframe
```
</summary>
<kbd>It would be <b>possible</b> if the Service Worker were started from the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code>, but that would create problems. See the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> alternative to starting the Service Worker for more information.</kbd>
</details>
<details><summary>

```nginx
Ping
from
popup
```
</summary>
<kbd>That is not possible without Chrome/Opera's bug.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
 Ping
 from
iframe
```
</summary>
<kbd>Ping the Service Worker every 29 seconds from inside the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> with <code>mitm.html</code>.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
'Ping
 from 
Shared
Worker
```
</summary>
<kbd>It is <b>possible</b>, but not necessary.</kbd>
</details>
<details><summary>

```nginx
Ping
from
popup
```
</summary>
<kbd>That is not possible without Chrome/Opera's bug.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```js
'Ping
 from 
Shared
Worker
```
</summary>
<kbd>Ping the Service Worker every 29 seconds from inside the Shared Worker.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```php
 Ping
 from
iframe
```
</summary>
<kbd>It is not possible. See the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> alternative to start the Service Worker for more information.</kbd>
</details>
<details><summary>

```nginx
Ping
from
popup
```
</summary>
<kbd>It is not possible without Chrome/Opera's bug.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
 Ping
 from
iframe
```
</summary>
<kbd>Ping the Service Worker every 29 seconds from inside the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> with <code>mitm.html</code>.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
'Ping
 from 
Shared
Worker
```
</summary>
<kbd>It is <b>possible</b>, but not necessary.</kbd>
</details>
<details><summary>

```nginx
Ping
from
popup
```
</summary>
<kbd>It is not possible without Chrome/Opera's bug.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
 Ping
 from
iframe
```
</summary>
<kbd>Ping the Service Worker every 4m30s from inside the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> with <code>mitm.html</code>.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
'Ping
 from 
Shared
Worker
```
</summary>
<kbd>It migh be <b>possible</b> (I don't think I tested this), but it is not necessary anyway.</kbd>
</details>
<details><summary>

```nginx
Ping
from
popup
```
</summary>
<kbd>It is <b>possible</b>, but not necessary.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
 Ping
 from
iframe
```
</summary>
<kbd>Ping the Service Worker every 4m30s from inside the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> with <code>mitm.html</code>.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
'Ping
 from 
Shared
Worker
```
</summary>
<kbd>It migh be <b>possible</b> (I don't think I tested this), but it is not necessary anyway.</kbd>
</details>
<details><summary>

```nginx
Ping
from
popup
```
</summary>
<kbd>It is <b>possible</b>, but not necessary.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```nginx
Ping
from
popup
  *
```
</summary>
<kbd>Ping the Service Worker every 4m30s from inside the closed popup with <code>mitm.html</code>.<br /><b>* Only works because of a bug.</b></kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
'Ping
 from 
Shared
Worker
```
</summary>
<kbd>It migh be <b>possible</b> (I don't think I tested this), but it is not necessary anyway.</kbd>
</details>
<details><summary>

```php
 Ping
 from
iframe
```
</summary>
<kbd>It is not possible. See the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> alternative to start the Service Worker for more information.</kbd>
</details>
</details></h6></td>
<td><h6><details><summary>

```php
 Ping
 from
iframe
```
</summary>
<kbd>Ping the Service Worker every 4m30s from inside the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> with <code>mitm.html</code>.</kbd>
<h4>Worse alternatives:</h4>
<details><summary>

```js
'Ping
 from 
Shared
Worker
```
</summary>
<kbd>It migh be <b>possible</b> (I don't think I tested this), but it is not necessary anyway.</kbd>
</details>
<details><summary>

```nginx
Ping
from
popup
```
</summary>
<kbd>It is <b>possible</b>, but not necessary.</kbd>
</details>
</details></h6></td></tr>
</table>

<h2><a name="changes-to-the-code">Changes to the code</a></h2>

Here I shall try to explain how the code was changed and why.

<h3><a name="streamsaverjs">StreamSaver.js</a></h3>

Three new detections were added alongside the secure context detection: <code>moz-extension:</code> protocol detection, background page detection and Firefox detection. Those are necessary for the branching that happens later.

The secure context detection now results in true for Web Extensions pages on Firefox that are on open tabs (the background page is excluded). This facilitates the branching, since Web Extension pages on tabs on Firefox should function in a very similar way to the secure pages.

A new parameter was added to indicate the URL of the <code>ping.html</code> page, used to connect to the Shared Worker needed in some contexts.

The code that was responsible for creating and sending a message to the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> on HTTPS pages was put in its own function, so it can be used in other places where the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> is needed (like for the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> with <code>ping.html</code>).

The code that was responsible for starting the download was merged into one function with the code that would open the popup. This function loads an URL in the correct place, which can be:
- a popup created by <code><a href="https://developer.mozilla.org/docs/Web/API/Window/open">window.open</a></code>;
- a popup created by <code><a href="https://developer.chrome.com/extensions/tabs#method-create">chrome.tabs.create</a></code>;
- <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code>.

This is where most of the branching actually happens.

A <code><a href="https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Math/random">Math.random()</a></code> value is now inserted into the hash of the <code>mitm.html</code> popup created by the background page of a Web Extension on Firefox. This hash serves two purposes:
1) Let the page know when it needs to load the Shared Worker;
2) Let the Shared Worker identify the right Service Worker instance when being messaged from the background page.

<h3><a name="swjs">sw.js</a></h3>

The message port received when registering a download is now stored as a property of the stream.
This is later used to notify StreamSaver.js when the Service Worker is used to start the download (this message triggers the removal of the download link's popup/<code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> when necessary).

The <code>'Mocking a download request'</code> message is removed so it won't interfere with the message that warns about the download starting.

After the Service Worker responds to a request for the download link it sends a message back to StreamSaver.js to notify about the download starting, as mentioned above.

<h3><a name="mitmhtml">mitm.html</a></h3>

For this to work on every scenario, a couple of changes were necessary.

Before this, the Service Worker was only registered after receiving the message to register a download. However, since the background page of Web Extensions on Firefox has no way to transfer ports to <code>mitm.html</code> directly, that would never happen.

Therefore, we would need the Shared Worker to be started right away, so the background page doesn't have to wait for a random amount of time until it is. Since it is also in the Shared Worker that the "pinging" happens, I have altered the code to start the Service Worker right away regardless of any messages (in the meantime, messages are stored for later processing). Once the Service Worker is registered, **only for the context of Web Extension's background pages** the Shared Worker is also started (only for this context; the hash is what tells this is the case).

This also has the benefit of only trying to register/recover the Service Worker once, insted of once per download.

The only other change is on the <code>setInterval</code> time: now it is 29 seconds if on Firefox, and 4.5 minutes everywhere else.

<h3><a name="pinghtml">ping.html</a></h3>

This is a new file created to be a bridge to the Shared Worker, mainly to support "pinging" the Service Worker from unsafe contexts or different processes.

It is a VERY simple code that only starts the Shared Worker and forwards any message it receives to it.

<h3><a name="pingjs">ping.js</a></h3>

This is a new file created to be the code of the Shared Worker.

It serves two purposes:
- Pinging the Service Worker to keep it alive;
- Transfering message ports to/from different processes to allow for the Web Extension's background pages on Firefox to communicate with the Service Worker directly.

<h2><a name="changes-to-functionality">Changes to functionality</a></h2>

Here I shall try to explain how the workings of the library were changed and why.

Any context not listed here should keep the previous functionality unchanged.

<h3><a name="download-from-iframe">Download from <code>&lt;iframe&gt;</code></a></h3>

Any place where the download was being started by navigating to the download link is now changed to load the download link inside the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> and remove the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> after the download starts.
This was done because navigating away from the page closed all open connections, and that sometimes affects other code not from StreamSaver.js.

<h3><a name="pinging-on-chromeopera">Pinging on Chrome/Opera</a></h3>

The pinging was happening every 29 seconds for Firefox compatibility, but now StreamSaver.js pings every 29 seconds only on Firefox, restoring the 4.5 minutes ping on Chrome/Opera.

<h3><a name="web-extensions-background-page-on-chromeopera">Web Extension's background page on Chrome/Opera</a></h3>

The download was started with a popup created with the <code><a href="https://developer.chrome.com/extensions/tabs#method-create">chrome.tabs.create</a></code> function.
Now it is started with <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code>.
This means no popups are ever opened on this context.

<h3><a name="http-pages-on-firefox">HTTP pages on Firefox</a></h3>

The Service Worker was getting killed after about 30 seconds, since nothing was keeping it alive.
Now, the Shared Worker is used to keep it alive for as long as necessary.

<h3><a name="web-extensions-on-firefox">Web Extensions on Firefox</a></h3>

Web Extensions on Firefox were not supported before, but now they are.
Web Extension's pages loaded on tabs are capable of starting downloads without opening a single popup, but background pages need to open two popups per download.
No user interaction is required, though.

<h2><a name="examples-for-testing">Examples for testing</a></h2>

I have created some example pages / Web Extensions to test all the changes in all contexts.

Each test has its own separate URL path for <code>mitm.html</code> and <code>sw.js</code>, so you can test them all simultaneously without one affecting the other.

<h3><a name="web-extensions-background-page-download">Web Extension's background page download</a></h3>

This is a simple Web Extension. The only thing it does is start a download from the background page, transfering one <code>'a'</code> letter per second during the course of one hour.

Download the extension and extract it: [https://texkiller.eu.org/StreamSaver/compatibility-back.zip](https://texkiller.eu.org/StreamSaver/compatibility-back.zip)

<table>
  <tr></tr>
  <tr>
    <th colspan="3">
      <h4>How to test it</h4>
    </th>
  </tr>
  <tr></tr>
  <tr>
    <th>
      <h5>Firefox v65+</h5>
    </th>
    <th>
      <h5>Chrome v52+</h5>
    </th>
    <th>
      <h5>Opera v39+</h5>
    </th>
  </tr>
  <tr></tr>
  <tr>
    <td>
      <ul>
        <li>Navigate to the URL <code>about:debugging#addons</code></li>
        <li>Make sure the "Enable add-on debugging" checkbox is marked</li>
        <li>Click on "Load Temporary Add-on...</li>
        <li>Select the file <code>manifest.json</code> inside the extracted folder</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>Navigate to the URL <code>chrome://extensions/</code></li>
        <li>Make sure the "Developer mode" checkbox is marked</li>
        <li>Click on "Load unpacked"</li>
        <li>Select the extracted folder</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>Navigate to the URL <code>opera://extensions/</code></li>
        <li>If you don't see the button labeled "Load unpacked extension...", click on "Developer mode"</li>
        <li>Click on "Load unpacked extension..."</li>
        <li>Select the extracted folder</li>
      </ul>
    </td>
  </tr>
  <tr></tr>
  <tr>
    <td colspan="3">
      <ul>
        <li>The sample extension should load and the download should be started automatically from the background page</li>
      </ul>
    </td>
  </tr>
</table>

<h3><a name="web-extensions-page-on-a-tab-download">Web Extension's page on a tab download</a></h3>

Another simple Web Extension. The only thing it does is open a page from the Web Extension into a tab with a button that, when pressed, starts a download transfering one <code>'a'</code> letter per second during the course of one hour.

Download the extension and extract it: [https://texkiller.eu.org/StreamSaver/compatibility-tab.zip](https://texkiller.eu.org/StreamSaver/compatibility-tab.zip)

<table>
  <tr></tr>
  <tr>
    <th colspan="3">
      <h4>How to test it</h4>
    </th>
  </tr>
  <tr></tr>
  <tr>
    <th>
      <h5>Firefox v65+</h5>
    </th>
    <th>
      <h5>Chrome v52+</h5>
    </th>
    <th>
      <h5>Opera v39+</h5>
    </th>
  </tr>
  <tr></tr>
  <tr>
    <td>
      <ul>
        <li>Navigate to the URL <code>about:debugging#addons</code></li>
        <li>Make sure the "Enable add-on debugging" checkbox is marked</li>
        <li>Click on "Load Temporary Add-on...</li>
        <li>Select the file <code>manifest.json</code> inside the extracted folder</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>Navigate to the URL <code>chrome://extensions/</code></li>
        <li>Make sure the "Developer mode" checkbox is marked</li>
        <li>Click on "Load unpacked"</li>
        <li>Select the extracted folder</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>Navigate to the URL <code>opera://extensions/</code></li>
        <li>If you don't see the button labeled "Load unpacked extension...", click on "Developer mode"</li>
        <li>Click on "Load unpacked extension..."</li>
        <li>Select the extracted folder</li>
      </ul>
    </td>
  </tr>
  <tr></tr>
  <tr>
    <td colspan="3">
      <ul>
        <li>The sample extension should load and a tab should show a page with a button</li>
        <li>Click the button to test</li>
      </ul>
    </td>
  </tr>
</table>

<h3><a name="web-page-on-http">Web page on HTTP</a></h3>

This is a simple web page on the HTTP protocol.
It shows a button that, when clicked, starts a download transfering one <code>'a'</code> letter per second during the course of one hour.

<table>
  <tr></tr>
  <tr>
    <th colspan="3">
      <h4>How to test it</h4>
    </th>
  </tr>
  <tr></tr>
  <tr>
    <th>
      <h5>Firefox v65+</h5>
    </th>
    <th>
      <h5>Chrome v52+</h5>
    </th>
    <th>
      <h5>Opera v39+</h5>
    </th>
  </tr>
  <tr></tr>
  <tr>
    <td colspan="3">
      <ul>
        <li>Navigate to the URL <code><a href="http://texkiller.eu.org/StreamSaver/compatibility/http/">http://texkiller.eu.org/StreamSaver/compatibility/http/</a></code></li>
        <li>Click the button to test</li>
      </ul>
    </td>
  </tr>
</table>

<h3><a name="web-page-on-https">Web page on HTTPS</a></h3>

This is a simple web page on the HTTPS protocol.
It shows a button that, when clicked, starts a download transfering one <code>'a'</code> letter per second during the course of one hour.

<table>
  <tr></tr>
  <tr>
    <th colspan="3">
      <h4>How to test it</h4>
    </th>
  </tr>
  <tr></tr>
  <tr>
    <th>
      <h5>Firefox v65+</h5>
    </th>
    <th>
      <h5>Chrome v52+</h5>
    </th>
    <th>
      <h5>Opera v39+</h5>
    </th>
  </tr>
  <tr></tr>
  <tr>
    <td colspan="3">
      <ul>
        <li>Navigate to the URL <code><a href="https://texkiller.eu.org/StreamSaver/compatibility/https/">https://texkiller.eu.org/StreamSaver/compatibility/https/</a></code></li>
        <li>Click the button to test</li>
      </ul>
    </td>
  </tr>
</table>

<h2><a name="things-to-look-forward-to">Things to look forward to</a></h2>

Here I shall mention changes in browsers that could open up new alternatives to the currently selected.

<h3><a name="transferable-streams">Transferable Streams</a></h3>

Explanation: [https://github.com/whatwg/streams/blob/master/transferable-streams-explainer.md](https://github.com/whatwg/streams/blob/master/transferable-streams-explainer.md)

Transferable Streams are expected to ease some of the library aspects, such as:

- There should no longer be a need to keep the Service Worker alive by pinging;
- There should no longer be a need to keep a shared message port between the page and the Service Worker;

A Chrome implementation seems to be underway, and should be on stable Chrome in one of the next versions.
I could find no bug on Bugzilla that addresses this matter specifically, though, so Firefox might take some time to get there.

<h3><a name="starting-service-worker-from-moz-extension">Starting Service Worker from <code>moz-extension://*</code></a></h3>

1344561 bug on Bugzilla: [https://bugzilla.mozilla.org/show_bug.cgi?id=1344561](https://bugzilla.mozilla.org/show_bug.cgi?id=1344561)

With this bug fixed, it should be always better to start the Service Worker from the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> instead of the <code><a href="https://developer.chrome.com/extensions/tabs#method-create">chrome.tabs.create</a></code> function on Web Extensions on Firefox.

This would mean background pages require one less popup there, and Web Extension's pages on tabs might be able to start the download by just navigating to the download URL.

The Shared Worker would be unnecessary on Web Extensions, with HTTP pages on Firefox being the only situation where it would be needed.

TexKiller commented:
Wow, this took a lot of time to make, and it is huge. I'm sorry for that. :/

jimmywarting commented:
<div align="center">   

![image](https://media1.tenor.com/images/891057fe32ead71265f447e559205711/tenor.gif)   

</div>

I'm sure it would have taken you a long time to make it. Need to test this, but not right now.

I tried this two on my Android using Brave
https://texkiller.eu.org/StreamSaver/compatibility/https/
http://texkiller.eu.org/StreamSaver/compatibility/http/

Did work just fine. But when i tried it out on my desktop  chrome browser i started to notice an error. It is not something wrong with the streamsaver or your code but it's the stream polyfill. You see, chrome have already implemented Transfarable, Transform & Writable streams so it's likely that the polyfill is not needed then, but it overwrites the global stream apis anyway. That causes a problem. readable streams are longer no longer transfarable even doe the variable `transfarableSupport` say that it's supported (but that was b/c streamsaver.js was loaded before the polyfill) so the streamsaver fails

It's also a problem with a mix of native+polyfilled version
You can't pipe a fetch response to a unnecessary polyfilled WritableStream
```js
rs = new ReadableStream() // polyfilled
ws = new WritableStream() // polyfilled

rs.pipeTo(ws) // ok (polyfilled <-> polyfilled)
rs = await new Response('').body // native
rs.pipeTo(ws) // fail (native <-> polyfilled)
```

Think we should do something about this problem, eg store the native TransformStream before the polyfill loads or something. The polyfill is still required by does that don't have WritableStream. Some browser only started out with ReadableStream

TexKiller commented:
Oh, I see you have the "Experimental Web Platform features" flag enabled to test the transferable streams on Chrome. I hadn't thought of testing that, so I didn't try to fix that problem.

I have now included a very small commit to fix that. It now works on my Google Chrome v73.0.3683.86 32bits with the "Experimental Web Platform features" flag enabled.

I have basically renamed and used the transferableSupport compatibility variable to store the compliant TransformStream class if transferable streams are supported.

TexKiller commented:
I've just noticed that the anchor links from the Table of Contents weren't working. They are now.

jimmywarting commented:
also notice that the stream polyfill also has a ponyfill, another idea that comes from some other packages are like 

```js
import fetch from 'node-fetch'
fetch.promise = bluebird
```
what if we made that for streamsaver?


```js
streamsaver.writableStream = require('web-stream-polyfill/ponyfill').WritableStream
```

Thoughts?

jimmywarting commented:
Yaa i agree
The default value would be 

Streamsaver.writablestream = window. Writablelestream || WebStreamPolyfill.Writablestream

jimmywarting commented:
If you could include it it would be great

TexKiller commented:
Done.

Also, all the <a href="#examples-for-testing">Examples for testing</a> have been updated to use the ponyfill instead of the polyfill.

TexKiller commented:
Do you want me to switch the polyfill for the ponyfill on the readme and on the example?

jimmywarting commented:
can leave the readme out until we make a new release on gh-pages and on npm and Tag it

TexKiller commented:
Sorry for the time without work, I have another project that is taking up all of my time, but it should be done in about two days and I'll get back to this.

jimmywarting commented:
It's okey. Don't feel pressured.

TexKiller commented:
Ok, so... Just found another problem.

Apparently, Chrome is also picky on security now. Turns out the <code><a href="https://developer.mozilla.org/docs/Web/HTML/Element/iframe">&lt;iframe&gt;</a></code> inside HTTP pages ignore Service Workers, and load the 404 page.

You can try it on the HTTP example on Chrome: it should never start the download. Make sure the old version isn't cached.

I think maybe we'll have to open the download on a popup there. On Firefox it seems to work in every context, and the other contexts seem to work fine on Chrome. I'll do some testing with the popup.

TexKiller commented:
And yet another problem:

It seems Firefox sometimes won't execute intervals in time (apparently due to load, I think), so the Service Worker was dying before the pinging happened. I'm thinking of changing the delay for the ping interval on Firefox to 10secs, so there is no chance it doesn't happen at least once in 30 seconds.

What do you think?

jimmywarting commented:
> Apparently, Chrome is also picky on security. Turns out the <iframe> inside HTTP pages ignores Service Workers, and loads the 404 page.

I did know that already, hence why i used a popup in the first place. the spec says if the top origin is not a secure context it should prevent any service worker from installing

TexKiller commented:
I knew we couldn't install Service Workers there, I just didn't know a Service Worker installed on a popup would be incapable of intercepting requests made there. It intercepts them fine on Firefox.

Anyway, what do you think about the delay on the ping interval on Firefox? That seems to be the only thing left to fix before it all works.

jimmywarting commented:
fine by me

jimmywarting commented:
Cool! All done?

Any breaking changes that require a major bump?

Summary of what have change?

TexKiller commented:
Well, I think it should all work now, but I would like to run all the tests again on all platforms to be sure. I'll be done with that in one hour, I think.

What has changed on this PR is all listed on <a href="#user-content-changes-to-the-code">Changes to the code</a> and <a href="#user-content-changes-to-functionality">Changes to functionality</a>. Just ignore the stuff about navigating to the download link, as that has been removed.

TexKiller commented:
Ok, tests done. Found a couple of peculiarities this time:

- Chrome with the Transferable Streams flag set won't start the download on HTTP only. Other protocols are fine with, and without Transferable Streams all protocols work. I don't know why that is happening now, it didn't happen before. Maybe the popup is messing with something.

- Version 58 of Opera won't allow the download popup to be opened on HTTP pages. Version 40 of Opera, however, does. I think they might have done a security patch somewhere in between, and that security patch is not on Chrome, so Chrome doesn't have this problem.

TexKiller commented:
I have updated the PR description to reflect the recent changes.

I have also changed all examples to include the code from <code>example.html</code>, as well as the simple button I was using. All the tests I ran were using my button, though, so there could be issues with some of the default examples that were not detected.

clivebor commented:
Hi @jimmywarting 

Thanks for this project, it's working great for us.

Will you be creating a new release with this compatibility overhaul?

jimmywarting commented:
Have some more things i have to deal with before making a new release
